webui/react/src/pages/ExperimentDetails/ExperimentCheckpoints.tsx

```diff
@@ -0,0 +1,287 @@
+
+  const dropDownOnTrigger = useCallback((checkpoint: CoreApiGenericCheckpoint) => {
+    return { [CheckpointAction.Register]: () => handleRegisterCheckpoint(checkpoint) };
+  }, [ handleRegisterCheckpoint ]);
+
+  const columns = useMemo(() => {
+    const actionRenderer = (_: string, record: CoreApiGenericCheckpoint): React.ReactNode => (
+      <ActionDropdown<CheckpointAction>
+        actionOrder={[
+          CheckpointAction.Register,
+        ]}
+        id={record.uuid}
+        kind="checkpoint"
+        onError={handleError}
+        onTrigger={dropDownOnTrigger(record)}
+      />
+    );
+
+    const checkpointRenderer = (_: string, record: CoreApiGenericCheckpoint): React.ReactNode => {
+      return (
+        <CheckpointModalTrigger
+          checkpoint={record}
+          experiment={experiment}
+          title={`Checkpoint ${record.uuid}`}
+        />
+      );
+    };
+
+    const newColumns = [ ...defaultColumns ].map((column) => {
+      column.sortOrder = null;
+      if (column.key === 'checkpoint') {
+        column.render = checkpointRenderer;
+      } else if (column.key === V1GetExperimentCheckpointsRequestSortBy.STATE) {
+        column.filterDropdown = stateFilterDropdown;
+        column.isFiltered = (settings) => !!(settings as Settings).state;
+        column.filters = ([ 'ACTIVE', 'COMPLETED', 'DELETED', 'ERROR' ] as CheckpointState[])
+          .map((value) => ({
+            text: <Badge state={value} type={BadgeType.State} />,
+            value,
+          }));
+      } else if (column.key === 'actions') {
+        column.render = actionRenderer;
+      }
+      if (column.key === settings.sortKey) {
+        column.sortOrder = settings.sortDesc ? 'descend' : 'ascend';
+      }
+      return column;
+    });
+
+    return newColumns;
+  }, [ dropDownOnTrigger, experiment, settings.sortDesc, settings.sortKey, stateFilterDropdown ]);
+
+  const handleTableChange = useCallback((tablePagination, tableFilters, tableSorter) => {
+    if (Array.isArray(tableSorter)) return;
+
+    const { columnKey, order } = tableSorter as SorterResult<CoreApiGenericCheckpoint>;
+    if (!columnKey || !columns.find((column) => column.key === columnKey)) return;
+
+    const newSettings = {
+      sortDesc: order === 'descend',
+      /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
+      sortKey: columnKey as any,
+      tableLimit: tablePagination.pageSize,
+      tableOffset: (tablePagination.current - 1) * tablePagination.pageSize,
+    };
+    const shouldPush = settings.tableOffset !== newSettings.tableOffset;
+    updateSettings(newSettings, shouldPush);
+  }, [ columns, settings.tableOffset, updateSettings ]);
+
+  const fetchExperimentCheckpoints = useCallback(async () => {
+    try {
+      const states = (settings.state ?? []).map((state) => (
+        encodeCheckpointState(state as CheckpointState)
+      ));
+      const response = await getExperimentCheckpoints(
```

Comment:
looks like this API is a couple years old and this is our first time using it in webui.

apparently it does a full scan over the checkpoints for the filter and paginates the result set in memory. might be worth doing some performance validation/asking around.. ?

I think it's probably not the worst offender, but worth mentioning at least.